storage/lmdb/src/remap.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for memory map reallocation
+
+use logging::log;
+use utils::tap_error_log::LogError;
+
+/// Represents LMDB memory map size
+#[derive(Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Debug)]
+pub struct MemSize(usize);
+
+impl MemSize {
+    /// Specify in the number of bytes
+    pub const fn from_bytes(bytes: usize) -> Self {
+        Self(bytes)
+    }
+
+    /// Specify in the number of kilobytes
+    pub const fn from_kilobytes(kilobytes: usize) -> Self {
+        Self::from_bytes(1024 * kilobytes)
+    }
+
+    /// Specify in the number of megabytes
+    pub const fn from_megabytes(megabytes: usize) -> Self {
+        Self::from_kilobytes(1024 * megabytes)
+    }
+
+    /// Convert to raw byte count
+    pub const fn as_bytes(self) -> usize {
+        self.0
+    }
+
+    pub fn div_ceil(self, rhs: Self) -> usize {
+        self.0 / rhs.0 + (self.0 % rhs.0 > 0) as usize
+    }
+}
+
+impl std::fmt::Display for MemSize {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}B", self.0)
+    }
+}
+
+impl std::ops::Add for MemSize {
+    type Output = MemSize;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Self(self.0 + rhs.0)
+    }
+}
+
+impl std::ops::Mul<MemSize> for usize {
+    type Output = MemSize;
+
+    fn mul(self, rhs: MemSize) -> Self::Output {
+        MemSize(self * rhs.0)
+    }
+}
+
+/// Token representing acquisition of the memory map resource
+#[derive(Debug)]
+pub struct MemMapToken();
+
+impl MemMapToken {
+    pub fn new() -> Self {
+        Self()
+    }
+}
+
+/// A proof of having acquired the memory map resource exclusively
+pub type ExclusiveMemMapToken<'a> = std::sync::RwLockWriteGuard<'a, MemMapToken>;
+
+/// Memory remapping procedure. Ensure at least `headroom` free space is available
+pub fn remap(
+    env: &lmdb::Environment,
+    _map_token: ExclusiveMemMapToken<'_>,
+    headroom: MemSize,
+) -> storage_core::Result<()> {
+    // Get page size
+    let page_size = {
```

Comment:
I get that this PR is supposed to be dead-simple, but this call is thread-safe... can't we have a way to call this and do the size check before acquiring the lock of the token, to avoid having to acquire multiple locks on every transaction?